### ISSUE TITLE
Gradient accumulation support?

### ISSUE THREAD

**Issue by andreped** [keras-team/tf-keras#107]
**Describe the feature and the current behavior/state:**

Gradient accumulation is extremely useful when working with large images/volumetric data, using low-end hardware, or training on multiple GPUs. For me, the most important feature is to be able to use larger batch sizes without exhausting memory.

Currently, there does not seem to be a straightforward way to use gradient accumulation in Keras.

**What I have tried:**

In TF1, we created a wrapper that can be used on any optimizer, which changes how and when the update should happen. I have tried to implement such a method in TF2, greatly inspired by the attempt by other developers at TF-addons, such as @fsx950223 and @stefan-falk https://github.com/tensorflow/addons/pull/2525. However, I have not managed to get expected behaviour (see [here](https://github.com/andreped/GradientAccumulator/issues/2) to see some of the experiments I performed, and [here](https://github.com/andreped/GradientAccumulator/blob/main/GradientAccumulator/accumulator.py#L7) for the optimizer wrapper implementation).

I therefore looked around for alternative solutions and found [this suggestion](https://stackoverflow.com/questions/66472201/gradient-accumulation-with-custom-model-fit-in-tf-keras/66524901#66524901) on stack overflow. I have expanded upon this idea and it seems to be working. After some thorough debugging and benchmarking, I have made a simple solution available in [this repo](https://github.com/andreped/GradientAccumulator), such that at least one simple solution for GA exists in TF2.

**Proposed solution:**

The idea is extremely simple. Overload the train_step method of the tf.keras.Model and add gradient accumulation support there. In the end, I have produced [a simple model wrapper](https://github.com/andreped/GradientAccumulator/blob/main/GradientAccumulator/GAModelWrapper.py), which does that for you, which you _ideally_ should be able to apply to _any_ tf.keras.Model to enable gradient accumulation, like so:
```
model = tf.keras.Model(...)
model = GAModelWrapper(n_gradients=k, inputs=model.input, outputs=model.output)
```

However, there is definitely some work left to be done to make it handle _all_ scenarios, but it seems to be working fine on the use cases I have tested until now.

**So what remains?**

Currently, I am unsure whether this is the best approach. Perhaps there is a better way of solving this. A challenge might be to get distributed training working with multiple GPUs. I believe that was the biggest obstacle with the optimizer wrapper solution.

Are there any devs working on adding gradient accumulation support in Keras?

**Are you willing to contribute it:** Yes

**Comment by IvanUkhov** [comment]
Perhaps something like this can be used as a workaround:

https://blog.ivanukhov.com/2024/01/31/gradient-accumulation.html

**Comment by andreped** [comment]
> Perhaps something like this can be used as a workaround:

This only addresses the Adam optimizer. If you want to wrap "any" optimizer, use this instead:
https://github.com/andreped/GradientAccumulator?tab=readme-ov-file#why

If you dont want to wrap the optimizer itself, you can also wrap the model. You can see how to do that from the link above.

There is also a thorough documentation on how to add gradient accumulation to an existing model pipeline here:
https://gradientaccumulator.readthedocs.io/en/latest/background/gradient_accumulation.html

---
NOTE: That in Keras 3, they seem to have made a solution for gradient accumulation now:
https://github.com/keras-team/keras/pull/18948#issuecomment-1858991628

It is a PoC and not feature complete, but a step in the right direction :]

**Comment by IvanUkhov** [comment]
The approach was illustrated on Adam, but it is not limited to that. The code is very general. It is a matter of replacing which class is inherited from.

The optimizer wrapper in GradientAccumulator, on the other hand, adds a lot of fragile complexity. It does not play well with the latest version of TensorFlow where the internals of optimizers have been redesigned. Moreover, it is limited to SGD according to the documentation. Please correct me if that is not the case.

https://gradientaccumulator.readthedocs.io/en/latest/examples/distributed_training.html

**Comment by andreped** [comment]
> It does not play well with the latest version of TensorFlow where the internals of optimizers have been redesigned. Moreover, it is limited to SGD according to the documentation. Please correct me if that is not the case.

Thats not the case. We have not upgraded the code itself to work with the latest 2.x versions, but I believe the Legacy Optimizer class is still there, which makes it work also for newer versions of TensorFlow. The implementation you showed is similar to what I made once in TF1, and that worked fine _for adam_, but even then I had an optimizer wrapper that worked well for both SGD, Adam, and other optimizers I normally used:
https://github.com/AICAN-Research/H2G-Net/blob/main/src/utils/accum_optimizers.py#L11

The optimizer wrapper implementation in `gradient-accumulator` should work with "any" commonly used optimizer, Adam included. But it is likely a better options to extend the optimizer wrapper, if you want to support more obscene optimizers.

What you are referring to is doing gradient accumulation in a distributed setting. I do not think there is **any** implementation in TF2 that works correctly in this scenario. Maybe Hugging Face has one, but the best is just to move to Keras 3 if that is of interest, IMO.

The implementation you showed does not work correctly in a distributed setting. You need more logic to handle how to handle replicas. That is not easy to solve, and why there does not seem to be a solution that does this properly. It is also not possible to do this in a model wrapper setting, as the syncronization from replicas results in an error. The optimizer wrapping is the only solution. **Optimizer wrapping is actually what Keras 3 does as well.**

But if you attempt to do GA where you split the model in two, across two GPUs, the optimizer wrapping solution should work. But if you want to split the batch across four GPUs, it does not. There is some logic missing and not trivial to solve, IMO.

**Comment by IvanUkhov** [comment]
> What you are referring to is doing gradient accumulation in a distributed setting. I do not think there is any implementation in TF2 that works correctly in this scenario. Maybe Hugging Face has one, but the best is just to move to Keras 3 if that is of interest, IMO.

Yes, that is my use case. Distributed training on several GPUs with the mirrored strategy.

> The implementation you showed does not work correctly in a distributed setting. You need more logic to handle how to handle replicas. That is not easy to solve, and why there does not seem to be a solution that does this properly. It is also not possible to do this in a model wrapper setting, as the syncronization from replicas results in an error. The optimizer wrapping is the only solution. Optimizer wrapping is actually what Keras 3 does as well.

That logic resides in the optimizer one inherits from, which was Adam in the example given. They are all distributed-training aware. The implementation works well. Do you have any specific concerns?

**Comment by andreped** [comment]
> That logic resides in the optimizer one inherits from, which was Adam in the example given. They are all distributed-training aware. The implementation works well. Do you have any specific concerns?

Well thats very interesting. It could be that we should try to not override most of the new Optimizer class for newer TF versions. That way, this might just magically work. My issue, is that I have gotten it mechanically to work before, using multiple GPUs in a distributed setting, but the main problem is that I do not get the expected GPU memory use across each GPU and I fail to see the same benefit of increasing the batch size by 4 by splitting across 4 GPUs, as I see for a single-GPU sequential setup.

But if you have managed to get it working for Adam, we could try to run benchmarks in Kaggle (2 GPUs available for free). And if that works, we could just rewrite the Optimizer wrapper to use the new Optimizer class, which may require **a lot** less code :]

Are you interested in posting that issue here:
https://github.com/andreped/GradientAccumulator/issues

And then we could maybe make a PR for it, @IvanUkhov, if we find that your solution works for newer TF versions.

**Comment by sachinprasadhs** [comment]
Thank you for reporting the issue. 
We're now prioritizing [Keras 3](https://keras.io/keras_3/) development, a multi-backend framework that offers significant advantages by supporting `JAX`, `TensorFlow`, `PyTorch` and `OpenVINO`. 
This means we won't be adding new features to `TF-Keras (Keras 2)`. 

`TensorFlow` users can benefit directly from `Keras 3`, which is included by default starting with version `2.16.1`.
 We encourage you to migrate to `Keras 3` and experience its enhanced capabilities. If you feel this specific feature is crucial for the `Keras` framework itself, please open a new issue at: https://github.com/keras-team/keras/issues

For more details and usage guide about Keras 3, refer: https://keras.io/getting_started/about/
